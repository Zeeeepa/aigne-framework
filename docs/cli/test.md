# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
