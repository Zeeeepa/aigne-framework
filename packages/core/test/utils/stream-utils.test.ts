import { expect, test } from "bun:test";
import { EventEmitter } from "node:stream";
import {
  agentResponseStreamToObject,
  arrayToAgentProcessAsyncGenerator,
  arrayToReadableStream,
  asyncGeneratorToReadableStream,
  mergeAgentResponseChunk,
  mergeReadableStreams,
  objectToAgentResponseStream,
  onAgentResponseStreamEnd,
  readAllString,
  readableStreamToArray,
  stringToAgentResponseStream,
  toReadableStream,
} from "@aigne/core/utils/stream-utils.js";

test("objectToAgentResponseStream should generate stream correctly", async () => {
  const stream = objectToAgentResponseStream({ foo: "foo", bar: "bar" });
  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("mergeAgentResponseChunk should merge correctly", async () => {
  // New field
  expect(mergeAgentResponseChunk({}, { delta: { text: { text: "hello" } } })).toEqual({
    text: "hello",
  });

  // Merge existing field
  expect(
    mergeAgentResponseChunk({ text: "hello" }, { delta: { text: { text: " world" } } }),
  ).toEqual({
    text: "hello world",
  });

  // Override existing text
  expect(
    mergeAgentResponseChunk({ text: "hello" }, { delta: { json: { text: "world" } } }),
  ).toEqual({
    text: "world",
  });
});

test("agentResponseStreamToObject should process asyncGenerator correctly", async () => {
  const result = await agentResponseStreamToObject<{
    text: string;
    foo: string;
  }>(
    arrayToAgentProcessAsyncGenerator(
      [
        { delta: { text: { text: "hello" } } },
        { delta: { text: { text: "," } } },
        { delta: { text: { text: " world" } } },
      ],
      {
        foo: "foo",
      },
    ),
  );
  expect(result).toEqual({
    text: "hello, world",
    foo: "foo",
  });
});

test("agentResponseStreamToObject should process readableStream correctly", async () => {
  const result = await agentResponseStreamToObject(
    arrayToReadableStream([
      { delta: { text: { text: "hello" } } },
      { delta: { text: { text: "," } } },
      { delta: { text: { text: " world" } } },
    ]),
  );
  expect(result).toEqual({
    text: "hello, world",
  });
});

test("asyncGeneratorToReadableStream should process readableStream correctly", async () => {
  const stream = asyncGeneratorToReadableStream(
    arrayToAgentProcessAsyncGenerator([{ delta: { json: { text: "hello" } } }], {
      text: "hello, world",
    }),
  );

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("arrayToAgentResponseStream should enqueue error", async () => {
  const stream = arrayToReadableStream([new Error("test error")]);
  const reader = stream.getReader();
  expect(reader.read()).rejects.toThrowError("test error");
});

test("arrayToAgentResponseStream should enqueue data", async () => {
  const stream = arrayToReadableStream([{ delta: { json: { text: "hello" } } }]);
  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("readableStreamToArray should collect chunks correctly", async () => {
  const stream = arrayToReadableStream([
    { delta: { text: { text: "hello" } } },
    { delta: { text: { text: " world" } } },
  ]);
  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("stringToAgentResponseStream should generate stream correctly", async () => {
  expect(
    readableStreamToArray(stringToAgentResponseStream("Hello, How can I assist you today?")),
  ).resolves.toMatchSnapshot();
});

test("stringToAgentResponseStream should generate stream with Chinese correctly", async () => {
  expect(
    readableStreamToArray(stringToAgentResponseStream("你好，我能帮你什么？", "custom_text")),
  ).resolves.toMatchSnapshot();
});

test("toReadableStream should generate stream correctly", async () => {
  const original = new EventEmitter() as unknown as NodeJS.ReadStream;

  const stream = readableStreamToArray(toReadableStream(original));

  original.emit("data", Buffer.from("Hello, world!"));
  original.emit("end");

  expect(await stream).toEqual([new Uint8Array(Buffer.from("Hello, world!"))]);
});

test("toReadableStream should throw error from input stream", async () => {
  const original = new EventEmitter() as unknown as NodeJS.ReadStream;

  const stream = readableStreamToArray(toReadableStream(original));

  original.emit("error", new Error("test error"));

  expect(stream).rejects.toThrowError("test error");
});

test("readAllString should read all string from ReadStream", async () => {
  const original = new EventEmitter() as unknown as NodeJS.ReadStream;

  const stream = readAllString(toReadableStream(original));

  original.emit("data", Buffer.from("Hello, world!"));
  original.emit("end");

  expect(await stream).toEqual("Hello, world!");
});

test("readAllString should read all string from ReadableStream", async () => {
  const stream = readAllString(
    new ReadableStream({
      start(controller) {
        controller.enqueue(new TextEncoder().encode("Hello, "));
        controller.enqueue(new TextEncoder().encode("world!"));
        controller.close();
      },
    }),
  );

  expect(await stream).toEqual("Hello, world!");
});

test("onAgentResponseStreamEnd should continue reading until end", async () => {
  const stream = onAgentResponseStreamEnd(
    arrayToReadableStream([
      { delta: { text: { text: "Hello " } } },
      { delta: {} },
      { delta: { text: { text: "world" } } },
    ]),
  );

  expect(await readableStreamToArray(stream)).toEqual([
    { delta: { text: { text: "Hello " } } },
    { delta: { text: { text: "world" } } },
  ]);
});

test("mergeReadableStreams should merge multiple streams", async () => {
  const stream1 = arrayToReadableStream([{ delta: { text: { text: "Hello" } } }]);
  const stream2 = arrayToReadableStream([{ delta: { text: { text: " world" } } }]);
  const mergedStream = mergeReadableStreams(stream1, stream2);

  expect(await readableStreamToArray(mergedStream)).toEqual([
    { delta: { text: { text: "Hello" } } },
    { delta: { text: { text: " world" } } },
  ]);
});

test("mergeReadableStreams should throw error from input stream", async () => {
  const stream1 = arrayToReadableStream([
    { delta: { text: { text: "Hello" } } },
    { delta: { text: { text: " world" } } },
    { delta: { text: { text: " SHOULD NOT MERGED" } } },
  ]);
  const stream2 = arrayToReadableStream([
    { delta: { text: { text: " 123" } } },
    new Error("test error"),
  ]);

  const mergedStream = mergeReadableStreams(stream1, stream2);
  expect(await readableStreamToArray(mergedStream, { catchError: true })).toMatchInlineSnapshot(`
    [
      {
        "delta": {
          "text": {
            "text": "Hello",
          },
        },
      },
      {
        "delta": {
          "text": {
            "text": " 123",
          },
        },
      },
      {
        "delta": {
          "text": {
            "text": " world",
          },
        },
      },
      [Error: test error],
    ]
  `);
});
