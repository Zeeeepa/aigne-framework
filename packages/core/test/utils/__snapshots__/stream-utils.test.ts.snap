// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`objectToAgentResponseStream should generate stream correctly 1`] = `
[
  {
    "delta": {
      "json": {
        "bar": "bar",
        "foo": "foo",
      },
    },
  },
]
`;

exports[`asyncGeneratorToReadableStream should process readableStream correctly 1`] = `
[
  {
    "delta": {
      "json": {
        "text": "hello",
      },
    },
  },
  {
    "delta": {
      "json": {
        "text": "hello, world",
      },
    },
  },
]
`;

exports[`arrayToAgentResponseStream should enqueue data 1`] = `
[
  {
    "delta": {
      "json": {
        "text": "hello",
      },
    },
  },
]
`;

exports[`readableStreamToArray should collect chunks correctly 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " world",
      },
    },
  },
]
`;

exports[`stringToAgentResponseStream should generate stream correctly 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "How",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "can",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "I",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "assist",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "you",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "today",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "?",
      },
    },
  },
]
`;

exports[`stringToAgentResponseStream should generate stream with Chinese correctly 1`] = `
[
  {
    "delta": {
      "text": {
        "custom_text": "你好",
      },
    },
  },
  {
    "delta": {
      "text": {
        "custom_text": "，",
      },
    },
  },
  {
    "delta": {
      "text": {
        "custom_text": "我能",
      },
    },
  },
  {
    "delta": {
      "text": {
        "custom_text": "帮",
      },
    },
  },
  {
    "delta": {
      "text": {
        "custom_text": "你",
      },
    },
  },
  {
    "delta": {
      "text": {
        "custom_text": "什么",
      },
    },
  },
  {
    "delta": {
      "text": {
        "custom_text": "？",
      },
    },
  },
]
`;
